File: 8kyu/(8 kyu) Square(n) Sum/(8 kyu) Square(n) Sum.js
// 1 Plain solution
function squareSum(numbers) { 
  let totalSum = 0;
  
  for (i=0; i < numbers.length; i++) {
    totalSum += Math.pow(numbers[i], 2);
  }
  
  return totalSum;
}

// 2 Optimized solution
function squareSum(numbers){
  let totalSum = 0;
  
  numbers.forEach(x =>
    totalSum +=  Math.pow(x, 2)
  );
  
  return totalSum;
}

// 3 Clever solution
function squareSum(numbers){
    return numbers
      .reduce(
        (acc, x) => acc + (x * x)
        , 0)
}

// 4 Coding golf
squareSum = n => n
  .reduce(
    (a, x) => a + x * x
    , 0
  );

File: 8kyu/(8 kyu) Square(n) Sum/(8 kyu) Square(n) Sum.dart
// 1 Simple solution
int squareSum(List numbers) {
  double sum = 0;

  for (var num in numbers) {
    sum += num * num;
  }

  return sum.toInt();
}

// 2 Optimized solution
int squareSum(List<int> numbers) {
  return numbers.fold(0, (sum, num) => sum + num * num).toInt();
}

// 3 Coding golf
int squareSum(List<int> n) => n.fold(0, (a, x) => a + x * x).toInt();
